### PR TITLE
Add full support for calls on z/OS and enable corresponding tests

### PR DESCRIPTION
**Pull request by fjeremic**
This PR contains a series of commits which enables full support for JIT <-> native calls on z/OS. The change appears rather large due to the amount of test code that was enhanced to validate the support. There was also other issues encountered during development which were fixed. Please see the commit descriptions for details on what each change does and why.

The "meat" of the support can be found in the two commits titled:

- "Create the XPLINK Function Descriptor for OMR compiled bodies on z/OS"
- "Use method symbol address to get the entry point in Tril and JitBuilder"

These two commits give us the support and are not very large changes all in all.

Issue: #4719

### CODE REVIEW

**Comment by fjeremic** [comment]
Leaving this in draft mode as I still need to validate that I haven't broken OpenJ9 on z/OS, which I most certainly have, but the fix will be simple as soon as I test it. In the mean while this should be ready for review. I'll proceed and self-review the change to note the most important design decisions which we can discuss here.

@genie-omr build all

File: compiler/control/CompileMethod.cpp [hunk 1]
```diff
@@ -379,7 +379,7 @@ compileMethodFromDetails(
          // not ready yet...
          //OMR::MethodMetaDataPOD *metaData = fe.createMethodMetaData(&compiler);
 
-         startPC = compiler.cg()->getCodeStart();
+         startPC = (uint8_t*)compiler.getMethodSymbol()->getMethodAddress();
```
Comment (fjeremic) [inline compiler/control/CompileMethod.cpp]: This was the main change in thinking here. See the commit body (ab6d6a497983599844cce11f5c1c870e9b160d8c) for a full description of why we made this change.

File: compiler/z/codegen/SystemLinkagezOS.cpp [hunk 2]
```diff
@@ -551,40 +551,61 @@ TR::S390zOSSystemLinkage::generateInstructionsForCall(TR::Node * callNode, TR::R
...
+
+          FunctionDescriptor* fd = reinterpret_cast<FunctionDescriptor*>(callSymbol->castToMethodSymbol()->getMethodAddress());
 
-       TR::SymbolReference *callSymRef = callNode->getSymbolReference();
-       TR::Symbol *callSymbol = callSymRef->getSymbol();
+          genLoadAddressConstant(cg(), callNode, reinterpret_cast<uintptrj_t>(fd), epReg);
+          generateRSInstruction(cg(), TR::InstOpCode::getLoadMultipleOpCode(), callNode, aeReg, epReg, generateS390MemoryReference(epReg, 0, cg()));
+
+          TR::Instruction* callInstr = new (cg()->trHeapMemory()) TR::S390RILInstruction(TR::InstOpCode::BRASL, callNode, raReg, fd->func, callSymRef, cg());
+          callInstr->setDependencyConditions(preDeps);
```
Comment (fjeremic) [inline compiler/z/codegen/SystemLinkagezOS.cpp]: Loads the function's address from the function descriptor and dispatches the call.

File: jitbuilder/control/Jit.cpp [hunk 3]
```diff
@@ -189,19 +189,7 @@ internal_compileMethodBuilder(TR::MethodBuilder *m, void **entry)
...
-      uint64_t environment;
-      void* func;
-   };
-
-   FunctionDescriptor* fd = new FunctionDescriptor();
-   fd->environment = 0;
-   fd->func = *entry;
-
-   *entry = (void*) fd;
-#elif defined(AIXPPC)
```
Comment (fjeremic) [inline jitbuilder/control/Jit.cpp]: Finally! This hack is now removed and we don't need to do anything special for z/OS. Once AIX gets support we can also remove the AIX piece as well.

File: jitbuilder/release/cpp/samples/Mandelbrot.cpp [hunk 4]
```diff
@@ -340,6 +339,12 @@ main(int argc, char *argv[])
 
    printf("Step 5: output result buffer\n");
    FILE *out = (argc == 3) ? fopen(argv[2], "wb") : stdout;
+
+   // The file we generate here is in the PPM image format which has a simple header describing how to interpret the
+   // buffer. This string needs to be encoded as ASCII in the file generated, however on platforms such as z/OS whose
+   // native encoding is not ASCII this is problematic.
+   //
+   // TODO (#4901): Use the port library to output the header bytes for PPM image file format
```
Comment (fjeremic) [inline jitbuilder/release/cpp/samples/Mandelbrot.cpp]: Linking this PR with #4901 so GitHub notes the link for future whenever we fix this.

**Comment by fjeremic** [comment]
Also suggesting @aviansie-ben for review as well as I cannot tag him on the GitHub interface.

File: compiler/control/CompileMethod.cpp (hunk 1 above)
Comment (fjeremic) [reply compiler/control/CompileMethod.cpp]: To be explicit this is a design decision which I'd like @mstoodle and @aviansie-ben to comment on, and perhaps @Leonardo2718 as well. It is unfortunate we don't have a public API, because in that case `getCodeStart` would likely be a private API internal to the compiler, and we would only provide one public API to extract the "pointer to a function" we compiled with OMR.

File: jitbuilder/control/Jit.cpp (hunk 3 above)
Comment (mstoodle) [reply jitbuilder/control/Jit.cpp]: Sweet!

File: compiler/codegen/OMRCodeGenPhase.cpp [hunk 5]
```diff
@@ -180,48 +180,44 @@ OMR::CodeGenPhase::performProcessRelocationsPhase(TR::CodeGenerator * cg, TR::Co
...
-        }
+   if (debug("dumpCodeSizes"))
+      {
+      diagnostic("%08d   %s\n", cg->getCodeLength(), comp->signature());
+      }
 
-     if (comp->getCurrentMethod() == NULL)
-        {
-        comp->getMethodSymbol()->setMethodAddress(cg->getBinaryBufferStart());
-        }
```
Comment (fjeremic) [inline compiler/codegen/OMRCodeGenPhase.cpp]: This piece of code was alarming to me. Does anyone know why this exists? And why is it guarded by a `getCurrentMethod` null check? When would this value ever be `NULL`?

File: compiler/codegen/OMRCodeGenPhase.cpp [hunk 6]
```diff
@@ -180,48 +180,44 @@ OMR::CodeGenPhase::performProcessRelocationsPhase(TR::CodeGenerator * cg, TR::Co
...
+      {
+      traceMsg(comp, "\n<relocatableDataCG>\n");
+      if (comp->getOption(TR_TraceRelocatableDataDetailsCG)) // verbose output
+         {
+         uint8_t * relocatableMethodCodeStart = (uint8_t *)comp->getRelocatableMethodCodeStart();
+         traceMsg(comp, "Code start = %8x, Method start pc = %x, Method start pc offset = 0x%x\n", relocatableMethodCodeStart, cg->getCodeStart(), cg->getCodeStart() - relocatableMethodCodeStart);
+         }
+      cg->getAheadOfTimeCompile()->dumpRelocationData();
+      traceMsg(comp, "</relocatableDataCG>\n");
+      }
```
Comment (mstoodle) [inline compiler/codegen/OMRCodeGenPhase.cpp]: any chance you could split these changes into a separate commit? they interfere with appreciating the setup you've made here for a much better solution to the FD issue.

Comment (fjeremic) [reply compiler/codegen/OMRCodeGenPhase.cpp]: The reason this formatting change was made is because the indentation in it's current state is very confusing [1], and at first sight one may thing the:

```
     if (comp->getCurrentMethod() == NULL)
        {
        comp->getMethodSymbol()->setMethodAddress(cg->getBinaryBufferStart());
        }
```

piece of code is guarded by the `if (cg->getAheadOfTimeCompile() &&` check, which is what I first thought, but upon a more careful inspection, the indentation was hiding this fact.

If you wish to view that commit without whitespace changes GitHub can help here:
https://github.com/eclipse/omr/pull/4907/commits/ab6d6a497983599844cce11f5c1c870e9b160d8c?w=1

[1] https://github.com/eclipse/omr/blob/31db81f3d84a431736ac7a9ff1b3980158fae258/compiler/codegen/OMRCodeGenPhase.cpp#L182-L202

Comment (fjeremic) [reply compiler/codegen/OMRCodeGenPhase.cpp]: On second thought, I think I agree. For history of commits it would be beneficial to split up formatting from actual changes.

I've split up the actual change into 6e902d5, where it belongs, and the formatting change into 
27349e8. Hope that helps!

**Comment by fjeremic** [comment]
@genie-omr build all

**Comment by fjeremic** [comment]
Just remembered my comment [1] to myself to revert some tests we disabled because of the lack of support for callouts. Now that we have them we can re-enable the tests.

[1] https://github.com/eclipse/omr/issues/4719#issuecomment-576394781

**Comment by fjeremic** [comment]
@genie-omr build all

**Comment by fjeremic** [comment]
> Leaving this in draft mode as I still need to validate that I haven't broken OpenJ9 on z/OS, which I most certainly have, but the fix will be simple as soon as I test it.

Much to my surprise I was wrong. OpenJ9 works fine with this change. The reason is because `generateInstructionsForCall` is a virtual method [1] and OpenJ9 overrides this in it's JNI linkage, which inherits the XPLINK system linkage on z/OS [2].

I've validated OpenJ9 works perfectly fine with this change, hence I'm marking this PR as ready for review/merging.

[1] https://github.com/eclipse/omr/blob/b03105e81b1b020900afbb2871cbd45aaaf8048c/compiler/z/codegen/SystemLinkage.hpp#L118-L120
[2] https://github.com/eclipse/openj9/blob/e8215634090734f30ba084a83089b58dbe579a3b/runtime/compiler/z/codegen/J9SystemLinkagezOS.cpp#L65-L218

**Comment by fjeremic** [comment]
I had to rebase to absorb the changes in #4821 to use the new `SKIP_ON_<ARCH>` helpers in LinkageTest.cpp and MinimalTest.cpp. Rest of the change is as before.

@genie-omr build all

**Comment by fjeremic** [comment]
I had to rebase again to fix conflicts introduced by removal of `uintptrj_t` in #4931. Launching builds again:

@genie-omr build all

**Comment by fjeremic** [comment]
@genie-omr build all

**Comment by fjeremic** [comment]
@genie-omr build zos

**Comment by fjeremic** [comment]
@mstoodle @0xdaryl @Leonardo2718 any takers for a committer for this one?

**Comment by fjeremic** [comment]
As per comment by @mstoodle (https://github.com/eclipse/omr/pull/4964#issuecomment-602879497) I've added the patch from #4909 to this PR in 2e7226c so we have test coverage. Launching builds again.

@genie-omr build all

**Comment by fjeremic** [comment]
@genie-omr build zos

**Comment by fjeremic** [comment]
Looks like there are issues on the following builds:
https://ci.eclipse.org/omr/job/PullRequest-linux_x86-64/2027/
https://ci.eclipse.org/omr/job/PullRequest-osx_x86-64/1581/

I'm going to back out 2e7226c from this PR and reopen #4909 for further investigation.

File: compiler/z/codegen/OMRLinkage.cpp [hunk 7]
```diff
@@ -2035,31 +2035,13 @@ OMR::Z::Linkage::buildArgs(TR::Node * callNode, TR::RegisterDependencyConditions
...
       }
 
-   bool isXPLinkToPureOSLinkageCall = false;
    if (!self()->isFirstParmAtFixedOffset())
       {
       stackOffset = argSize;
       }
    else
       {
       stackOffset = self()->getOffsetToFirstParm();
```
Comment (Leonardo2718) [inline compiler/z/codegen/OMRLinkage.cpp]: Would it be worth simplifying this to

```c++
stackOffset = self()->isFirstParmAtFixedOffset() ? self()->getOffsetToFirstParm() : argSize;
```

File: fvtest/compilertriltest/LinkageTest.cpp [hunk 8]
```diff
@@ -61,8 +61,10 @@ TYPED_TEST(LinkageTest, InvalidLinkageTest) {
     SKIP_ON_PPC(MissingImplementation) << "Test is skipped on POWER because calls are not currently supported (see issue #1645)";
     SKIP_ON_PPC64(MissingImplementation) << "Test is skipped on POWER 64 because calls are not currently supported (see issue #1645)";
     SKIP_ON_PPC64LE(MissingImplementation) << "Test is skipped on POWER 64le because calls are not currently supported (see issue #1645)";
-    SKIP_ON_S390(MissingImplementation) << "Test is skipped on S390 because calls are not currently supported (see issue #1645)";
-    SKIP_ON_S390X(MissingImplementation) << "Test is skipped on S390x because calls are not currently supported (see issue #1645)";
+    if ("z/OS" != omrsysinfo_get_OS_type()) {
+       SKIP_ON_S390(MissingImplementation) << "Test is skipped on S390 because calls are not currently supported (see issue #1645)";
+       SKIP_ON_S390X(MissingImplementation) << "Test is skipped on S390x because calls are not currently supported (see issue #1645)";
+    }
```
Comment (Leonardo2718) [inline fvtest/compilertriltest/LinkageTest.cpp]: It seems unfortunate to have to use an if statement. Would it be worth implementing a macro like:

```c++
SKIP_ON_S390X_ZOS(MissingImplementation) << "...";
```

File: compiler/control/CompileMethod.cpp (hunk 1 above)
Comment (Leonardo2718) [reply compiler/control/CompileMethod.cpp]: This approach seems fine to me. 

I agree that having a single public API to extract the function pointer would be ideal. But I don't think that's a problem that needs to be solved here. The approach you've taken is reasonable given what we currently have, IMHO.

File: fvtest/compilertriltest/LinkageTest.cpp [hunk 9]
```diff
@@ -411,26 +590,221 @@ TYPED_TEST(LinkageTest, SystemLinkageParameterPassingFiveArgToStackUser) {
...
+T return6thArgumentFromMixedTypes(float a, int32_t b, double c, int32_t d, int64_t e, T f) { return f; }
+
+template <typename T>
+T return7thArgumentFromMixedTypes(double a, int32_t b, double c, int32_t d, int32_t e, int64_t f, T g) { return g; }
+
+template <typename T>
+T return8thArgumentFromMixedTypes(int64_t a, int64_t b, float c, double d, int32_t e, int64_t f, int32_t g, T h) { return h; }
+
+template <typename T>
+T return9thArgumentFromMixedTypes(double a, int32_t b, float c, int32_t d, int32_t e, float f, int64_t g, int32_t h, T i) { return i; }
```
[the remark this answers is not in the document]
Comment (fjeremic) [reply fvtest/compilertriltest/LinkageTest.cpp]: It's completely random.

**Comment by fjeremic** [comment]
@Leonardo2718 thanks for the review. I still want to address your earlier points. I think there is no better time do implement such work than now. I'll have fixes for those in the next day. Apologies for the slow progress.

**Comment by Leonardo2718** [comment]
> Apologies for the slow progress.

Well, if we're going to apologize for slow progress, I should apologies for taking so long to review. :sweat_smile:

**Comment by fjeremic** [comment]
@genie-omr build all

**Comment by fjeremic** [comment]
@genie-omr build all

**Comment by fjeremic** [comment]
@genie-omr build all

File: compiler/z/codegen/OMRLinkage.cpp (hunk 7 above)
Comment (fjeremic) [reply compiler/z/codegen/OMRLinkage.cpp]: Fixed in f5f00a4

File: fvtest/compilertriltest/LinkageTest.cpp [hunk 10]
```diff
@@ -488,29 +919,1744 @@ TEST_F(LinkageWithMixedTypesTest, SystemLinkageParameterPassingFiveArgWithMixedT
...
+       "  (block"
+       "    (%sreturn"
+       "      (%scall address=0x%jX args=[%s] linkage=system"
+       "        (%sload parm=0) ) ) ) )",
+
+       TypeToString<TypeParam>::type, // return
+       TypeToString<TypeParam>::type, // args
+       TypeToString<TypeParam>::prefix, // return
+       TypeToString<TypeParam>::prefix, //call
+       reinterpret_cast<uintmax_t>(static_cast<TypeParam(*)(TypeParam)>(callee_entry_point)), // address
```
[the remark this answers is not in the document]
Comment (fjeremic) [reply fvtest/compilertriltest/LinkageTest.cpp]: You're right. All the builds passed. Fixed in 11c8008

File: fvtest/compilertriltest/LinkageTest.cpp (hunk 8 above)
Comment (fjeremic) [reply fvtest/compilertriltest/LinkageTest.cpp]: Unfortunate indeed. No better time to fix it than now though. Added new macros and made use of them in 13a5b43

**Comment by fjeremic** [comment]
All review comments addressed. @Leonardo2718 do you want to be a committer for this one?

File: fvtest/compilertriltest/LinkageTest.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Leonardo2718) [reply fvtest/compilertriltest/LinkageTest.cpp]: If it worked in this one location, I would think that it should also work for every other test that uses `callee_entry_point`.

[the remark this answers is not in the document]
Comment (fjeremic) [reply fvtest/compilertriltest/LinkageTest.cpp]: Right. I've force pushed the fix to cover all uses in fa96327

**Comment by Leonardo2718** [comment]
@genie-omr build all

I want to merge this tomorrow by EOD. If anyone still wants to review this change, please speak up before then.

**Comment by fjeremic** [comment]
@genie-omr build zos

**Comment by Leonardo2718** [comment]
Times up! Merging now.
